provider/pkg/resources/cfn_custom_resource.go

```diff
@@ -0,0 +1,681 @@
+	var typedInputs CfnCustomResourceInputs
+	_, err := resourcex.Unmarshal(&typedInputs, inputs, resourcex.UnmarshalOptions{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal inputs: %w", err)
+	}
+
+	var failures []ValidationFailure
+
+	if !lambdaFunctionArnRegex.MatchString(typedInputs.ServiceToken) {
+		// TODO create a GitHub issue for this and link it
+		failures = append(failures, ValidationFailure{
+			Path:   "serviceToken",
+			Reason: "serviceToken must be a valid Lambda function ARN.",
+		})
+	}
+
+	if typedInputs.StackID == nil {
+		// if the stack ID is not provided, we use the pulumi stack ID as the stack ID
+		inputs[resource.PropertyKey("stackId")] = resource.NewStringProperty(urn.Stack().String())
+	}
+
+	if typedInputs.CustomResourceProperties != nil {
+		stringifiedCustomResourceProperties := naming.ToStringifiedMap(typedInputs.CustomResourceProperties)
+		inputs[resource.PropertyKey("customResourceProperties")] = resource.PropertyValue{V: resource.NewPropertyMapFromMap(stringifiedCustomResourceProperties)}
+	}
+
+	return inputs, failures, nil
+}
+
+// Create creates the Custom Resource by invoking the Lambda function with the CREATE request type.
+// Returns the physical resource ID and outputs if the creation is successful, otherwise an error.
+//
+// Creation Flow:
+//
+//   ┌─────────┐   ┌─────────────┐   ┌───────────────┐   ┌──────────┐
+//   │ Create  │   │Generate S3  │   │ Invoke Lambda │   │ Wait for │
+//   │ Request ├──►│Presigned URL├──►│ with CREATE   ├──►│ Response │
+//   └─────────┘   └─────────────┘   │ RequestType   │   └────┬─────┘
+//                                   └───────────────┘        │
+//                                                            ▼
+//                                                    ┌───────────────┐
+//                                                    │Return Physical│
+//                                                    │Resource ID &  │
+//                                                    │Outputs        │
+//                                                    └───────────────┘
+//
+// The Create operation:
+// 1. Generates a presigned S3 URL for response collection
+// 2. Constructs a CloudFormation CREATE event with:
+//    - Unique RequestID (UUID)
+//    - ResponseURL (presigned S3 URL)
+//    - ResourceType from input
+//    - LogicalResourceId from Pulumi URN
+//    - Custom properties from input
+// 3. Invokes the Lambda function asynchronously
+// 4. Waits for response in S3 bucket
+// 5. Processes response:
+//    - On success: Returns PhysicalResourceId and properties
+//    - On failure: Returns error with reason
+//    - Handles `NoEcho` for sensitive data
+func (c *cfnCustomResource) Create(ctx context.Context, urn urn.URN, inputs resource.PropertyMap, timeout time.Duration) (*string, resource.PropertyMap, error) {
```

Comment:
I was wondering if this is ever called in Preview, but looks like the provider is SupportsPreview: false so we're good. It never is! No unknowns then.